src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,609 @@
+    """Compress the cloud image.
+
+    Args:
+        image: The image to compress.
+        output: The desired image output path.
+
+    Raises:
+        ImageCompressError: If there was something wrong compressing the image.
+    """
+    try:
+        subprocess.run(  # nosec: B603
+            ["/usr/bin/virt-sparsify", "--compress", str(image), str(output)],
+            check=True,
+            timeout=60 * 10,
+        )
+    except subprocess.CalledProcessError as exc:
+        raise ImageCompressError from exc
+
+
+IMAGE_DEFAULT_APT_PACKAGES = [
+    "docker.io",
+    "npm",
+    "python3-pip",
+    "shellcheck",
+    "jq",
+    "wget",
+    "unzip",
+    "gh",
+]
+
+
+@dataclasses.dataclass
+class BuildImageConfig:
+    """Configuration for building the image.
+
+    Attributes:
+        arch: The CPU architecture to build the image for.
+        base_image: The ubuntu image to use as build base.
+        output: The path to write final image to.
+    """
+
+    arch: Arch
+    base_image: BaseImage
+    output: Path
+
+
+class BuildImageError(Exception):
+    """Represents an error while building the image."""
+
+
+def build_image(config: BuildImageConfig) -> None:
+    """Build and save the image locally.
+
+    Args:
+        config: The configuration values to build the image with.
+
+    Raises:
+        BuildImageError: If there was an error building the image.
+    """
+    logger.info("Clean build state.")
+    with redirect_stdout(sys.stderr):
+        _clean_build_state()
+        try:
+            logger.info("Downloading cloud image.")
+            cloud_image_path = _download_cloud_image(
+                arch=config.arch, base_image=config.base_image
+            )
+            logger.info("Resizing cloud image.")
+            _resize_cloud_img(cloud_image_path=cloud_image_path)
+            logger.info("Mounting network block device.")
+            _mount_image_to_network_block_device(cloud_image_path=cloud_image_path)
+            logger.info("Replacing resolv.conf.")
+            _replace_mounted_resolv_conf()
+            logger.info("Resizing partitions.")
+            _resize_mount_partitions()
+        except (CloudImageDownloadError, ImageMountError, ResizePartitionError) as exc:
+            raise BuildImageError from exc
+
+        try:
+            logger.info("Setting up chroot environment.")
+            with ChrootContextManager(IMAGE_MOUNT_DIR):
+                # operator_libs_linux apt package uses dpkg -l and that does not work well with
+                # chroot env, hence use subprocess run.
+                subprocess.run(
+                    ["/usr/bin/apt-get", "update", "-y"],
+                    check=True,
+                    timeout=60 * 10,
+                    env={"DEBIAN_FRONTEND": "noninteractive"},
+                )  # nosec: B603
+                subprocess.run(  # nosec: B603
+                    ["/usr/bin/apt-get", "install", "-y", *IMAGE_DEFAULT_APT_PACKAGES],
```

Comment:
Shouldn't we use `--no-install-recommends` ?

Reply:
Updated, thank you!